krux/parser.py

```diff
@@ -0,0 +1,263 @@
+    )
+    group.add_argument(
+        '--stats-port',
+        default=DEFAULT_STATSD_PORT,
+        help='Statsd port to send statistics to. (default: %(default)s)'
+    )
+    group.add_argument(
+        '--stats-environment',
+        default=DEFAULT_STATSD_ENV,
+        help='Statsd environment. (default: %(default)s)'
+    )
+
+    return parser
+
+
+def add_lockfile_args(parser):
+    group = get_group(parser=parser, group_name='lockfile', prefix=False)
+
+    group.add_argument(
+        '--lock-dir',
+        default=DEFAULT_LOCK_DIR,
+        help='Dir where lock files are stored (default: %(default)s)'
+    )
+
+    return parser
+
+
+def get_group(parser, group_name, prefix=None):
+    """
+    Return an argument group based on the group name.
+
+    This will return an existing group if it exists, or create a new one.
+
+    :argument parser: :py:class:`python3:argparse.ArgumentParser` to
+                      add/retrieve the group to/from.
+
+    :argument group_name: Name of the argument group to be created/returned.
+    """
+
+    # We don't want to add an additional group if there's already a 'logging'
+    # group. Sadly, ArgumentParser doesn't provide an API for this, so we have
+    # to be naughty and access a private variable.
+    groups = [g.title for g in parser._action_groups]
+
+    if group_name in groups:
+        group = parser._action_groups[groups.index(group_name)]
+    else:
+        group = parser.add_argument_group(title=group_name, prefix=prefix)
+
+    return group
+
+
+class KruxParser(Wrapper):
+    def add_argument_group(self, prefix, *args, **kwargs):
+        group = KruxGroup(
+            wrapped=_ArgumentGroup(self, *args, **kwargs), prefix=prefix, logger=self._logger, stats=self._stats,
+        )
+        self._wrapped._action_groups.append(group)
+        return group
+
+
+class KruxGroup(Wrapper):
+    def __init__(self, prefix=None, *args, **kwargs):
+        """
+        Creates a wrapper around argparse._ArgumentGroup that handles some help doc automation as well as environment
+        variable fall back
+
+        :param prefix: Prefix to use for the environment variables. If set to None, uses the title
+                       of the _ArgumentGroup that this is wrapping. If set to False, does not add any prefix
+        :type prefix: str | bool
+        :param args: Ordered arguments passed directly to krux.wrapper.Wrapper.__init__()
+        :type args: list
+        :param kwargs: Keyword arguments passed directly to krux.wrapper.Wrapper.__init__()
+        :type kwargs: dict
+        """
+        # Call to the superclass to bootstrap.
+        super(KruxGroup, self).__init__(*args, **kwargs)
+
+        if prefix is None:
+            self._env_prefix = self._wrapped.title + '_'
+        elif prefix is False:
+            self._env_prefix = ''
+        else:
+            self._env_prefix = str(prefix) + '_'
+
+    def add_argument(self, *args, **kwargs):
```

Comment:
It hadn’t occurred to me to intercept with our environment variables at the `add_argument()` stage instead of the parse_args() stage!

Reply:
This was the break-through that I thought of yesterday. :)